### DISCUSSION TITLE
`semanticCommitType` when same package is both a dependency and devDependency in a monorepo

### DISCUSSION THREAD

**Discussion by travi** [renovatebot/renovate#10908] [Suggest an Idea]
## Unexpected Behavior

i normally prefer to avoid monorepos, so i haven't run into this before, but i'm currently working in a lerna monorepo and have noticed cases where the [semanticCommitType](https://docs.renovatebot.com/configuration-options/#semanticcommittype) is correct for some consumers in a PR but incorrect for other consumers in the same PR. this is because that package is consumed as a devDependency by some of the consumers and as a prod dependency in the others.

the `chore` commit type is being used by renovate in this case, which does seem like the safe option. when used in combination with `semantic-release-monorepo`, the `chore` prevents worthless extra versions of the packages that consume the dependency as a devDependency from being released. however, that also prevents new versions of the packages that consume the dependency as a prod dependency (which would normally use a `fix` commit type) from being released when they normally would be. 

## Desired Behavior

my goal would be to have two groups:
* one that groups the consumers that use the dependency as a prod dependency, using the `fix` commit type
* one that groups the consumers that use the dependency as a devDependency, using the `chore` commit type

i could see two potential options:
* two separate PRs
* two separate commits in the same PR, with care applied so that automerging does not squash into a single commit

## Current configurability

is it possible to separate these groups with existing config options? what would it take to enable this type of grouping?

**Comment by rarkins** [comment]
When you have multiple/conflicting semantic commit config within the same branch then it is kind of "undefined behaviour" (the branch will use the values from the first upgrade in the list) so if you're seeing `chore` it's out of chance rather than a decision. Renovate doesn't really "understand" semantic commits or which are important or not - it treats them as strings.

If you want dev dependencies to be separated from production dependencies then you need to do that via branch naming. `additionalBranchPrefix` is maybe an option, e.g. through a packageRules matching against matchDepTypes=devDependencies and additionalBranchPrefix=dev-. We currently don't do one-commit-per-dependency and it would complicate some of our other concepts so it's not in our plan.

**Reply by travi** [reply]
## solution to my original question

thank you! that did the trick, mostly.

i already had `packageRules` for the following:
```json
    {
      "packagePatterns": [
        "*"
      ],
      "semanticCommitType": "chore"
    },
    {
      "depTypeList": [
        "dependencies",
        "peerDependencies"
      ],
      "semanticCommitType": "fix"
    }
```
and made the following addition:
```diff
    {
      "packagePatterns": [
        "*"
      ],
      "semanticCommitType": "chore"
    },
    {
      "depTypeList": [
        "dependencies",
        "peerDependencies"
      ],
      "semanticCommitType": "fix",
+      "additionalBranchPrefix": "prod-"
    }
```

that resulted in the the dev-dependency updates being left in the PR with the `chore` commit-type and a new PR was created with a `fix` commit type (might have happened the other way around, but the were split up as desired).

also, this was helpful. i knew that the semantic types arent really understood by renovate, but wasn't sure how things worked out in a case like this. updating my mental model a bit further does help me think through how to make these adjustments:

> Renovate doesn't really "understand" semantic commits or which are important or not - it treats them as strings.

## additional complexity

however, this made me realize that my situation was a little more complex than i'd realized at first due to having a peer-dependency defined for the same package (satisfying the peer-dep requirement for `npm ls` is actually the reason it is defined as a dev-dependency in the first place). 

as a result, the `fix` PR included the lockfile update for the package that defined the target package as a peer. the `chore` PR also updated the lockfile due to the dev-dependency udpate, so as long as that PR is merged first, we get back (after the rebase of the `fix` PR) to the state i'd like to be in the `fix` PR originally. ideally, i'd like all updates that only change a lockfile to be `chore`, regardless of dependency-type, but i havent found the time to dig in further than opening https://github.com/renovatebot/renovate/issues/6791 a while back. it feels like this introduces a scenario we didnt consider in the `in-range` potential solution there since the lockfile update would be out of range, but updated for the dev-dependency.

## reason for using `fix` for `peerDependencies`

as you can see in my rules above, i have `peerDependencies` marked `semanticCommitType: fix`. this is because i want a patch version to be released when the compatible peer versions get widened. i imagine there is a more specific way to accomplish this, but that config has accomplished my goal so far. 

## next steps

anyway, there are some details for me to think on a bit now that we solved my original question. the update i made as a result of this does unblock us and gives us a workable solution to move forward with for now. we just need to be careful about the order we merge the two PRs (`chore` first).

this gives me a better starting point to experiment a bit with possible configs to balance the three dependency types. the overlap of peer and dev types feels complicated enough that i will need to try out some options with a real repo to make sure i understand all of the scenarios. i'll try to report back here if i come up with a config that does what i'm after.
